rust/kernel/net/mod.rs

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Network subsystem in rust for the linux kernel.
+
+use core::mem;
+
+pub mod device;
+pub mod ethtool;
+pub mod netlink;
+pub mod rtnl;
+pub mod skbuff;
+
+#[doc(inline)]
+pub use module::rtnl_link_ops;
+
+/// Determine if the Ethernet address is a multicast.
+///
+/// Return true if the address is a multicast address.
+/// By definition the broadcast address is also a multicast address.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_multicast_ether_addr(addr: *const u8) -> bool {
+    let a: u32 = *(addr as *const u32);
+
+    if cfg!(target_endian = "big") {
+        (0x01 & (a >> (((mem::size_of::<u32>() as u32) * 8) - 8))) != 0
+    } else {
+        (0x01 & a) != 0
+    }
+}
+
+/// Determine if the Ethernet address is a multicast.
+///
+/// Return true if the address is a multicast address.
+/// By definition the broadcast address is also a multicast address.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+#[cfg(not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS))]
+pub unsafe fn is_multicast_ether_addr(addr: *const u8) -> bool {
+    let a: u16 = *(addr as *const u16);
+
+    if cfg!(target_endian = "big") {
+        (0x01 & (a >> (((mem::size_of::<u16>() as u16) * 8) - 8))) != 0
+    } else {
+        (0x01 & a) != 0
+    }
+}
+
+/// Determine if give Ethernet address is all zeros.
+/// Return true if the address is all zeroes.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// Please note: addr must be aligned to u16.
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_zero_ether_addr(addr: *const u8) -> bool {
+    *(addr as *const u32) | (*((addr as usize + 4) as *const u16) as u32) == 0
+}
+
+/// Determine if give Ethernet address is all zeros.
+/// Return true if the address is all zeroes.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// Please note: addr must be aligned to u16.
+#[cfg(not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS))]
+pub unsafe fn is_zero_ether_addr(addr: *const u8) -> bool {
+    *(addr as *const u16)
+        | *((addr as usize + 2) as *const u16)
+        | *((addr as usize + 4) as *const u16)
+        == 0
```

Comment:
I was thinking that
```suggestion
    *(addr as *const [u16; 3]) == [0; 3]
```
would be nicer to read. When I tested if LLVM would be able to optimize it I noticed the following: On x86_64 and aarch64 the current version uses 3 loads while the suggested version does 2 loads just like the `CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS` version. On armv7 this results in a branch + load from a constant pool for the suggested version while version uses 3 loads and no constant pool.

Because of this I think this suggestion should be used for `CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS` to avoid UB at no perf cost, but `not(`CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)` should remain the same.

Reply:
Would it maybe make sense for all of these methods to take a `&[u8; 6]` and be completely safe to call or is there an inherent reason we need to treat these values as `u16`s?

Reply:
For `CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS` there would be no change. for `not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)` that would require 6 loads (one for each u8) instead of 3 (one for each u16).